File: scratch/4sat-link-emu-multi-dynamic-p2p-graph.cc
#include "ns3/core-module.h"
#include "ns3/network-module.h"
#include "ns3/point-to-point-module.h"
#include "ns3/fd-net-device-module.h"
#include "ns3/error-model.h"
#include "ns3/ethernet-header.h"
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <map>

using namespace ns3;

void KeepAliveDummyEvent () {}

constexpr int numNodes = 4;   

// --- Full Mesh Tracking Matrices ---
// meshDevices[SrcNode][DstNode] gives the exact PointToPoint device transmitting FROM Src TO Dst.
Ptr<PointToPointNetDevice> meshDevices[numNodes][numNodes];   
Ptr<NetDevice> g_fdDev[numNodes];

// Global MAC Table for distributed switching
std::map<Mac48Address, uint32_t> macToNodeMap;

// --- 1. Custom Encapsulation Tunnel ---
void SendOverP2PTunnel (Ptr<NetDevice> dev, Ptr<const Packet> packet, uint16_t protocol, const Address &src, const Address &dst)
{
  Ptr<Packet> pktCopy = packet->Copy();
  EthernetHeader eth;
  eth.SetSource (Mac48Address::ConvertFrom (src));
  eth.SetDestination (Mac48Address::ConvertFrom (dst));
  eth.SetLengthType (protocol);
  
  pktCopy->AddHeader (eth);
  dev->Send (pktCopy, dev->GetBroadcast (), 0x0800); 
}

// --- 2. Distributed Switch Ingress (From Physical Raspberry Pi) ---
bool Ingress_From_Vlan (uint32_t myNodeID, Ptr<NetDevice> rxDevice, Ptr<const Packet> packet, uint16_t protocol,
                         const Address &src, const Address &dst, NetDevice::PacketType packetType)
{
  Mac48Address srcMac = Mac48Address::ConvertFrom(src);
  Mac48Address dstMac = Mac48Address::ConvertFrom(dst);

  // Learn where this Pi's MAC address lives
  if (!srcMac.IsBroadcast())
    {
      macToNodeMap[srcMac] = myNodeID;
    }

  // If Broadcast/Multicast (like ARP or OSPF), flood to all OTHER nodes in the mesh
  if (dstMac.IsBroadcast() || dstMac.IsGroup())
    {
      for (uint32_t i = 0; i < numNodes; ++i)
        {
          if (i != myNodeID && meshDevices[myNodeID][i] != nullptr)
            {
              SendOverP2PTunnel (meshDevices[myNodeID][i], packet, protocol, src, dst);
            }
        }
    }
  else // Unicast
    {
      if (macToNodeMap.find(dstMac) != macToNodeMap.end())
        {
          uint32_t targetNode = macToNodeMap[dstMac];
          // Forward directly down the dedicated mesh wire to the destination
          if (targetNode != myNodeID && meshDevices[myNodeID][targetNode] != nullptr)
            {
              SendOverP2PTunnel (meshDevices[myNodeID][targetNode], packet, protocol, src, dst);
            }
        }
      else
        {
          // Unknown unicast: Flood to all other nodes until MAC is learned
          for (uint32_t i = 0; i < numNodes; ++i)
            {
              if (i != myNodeID && meshDevices[myNodeID][i] != nullptr)
                {
                  SendOverP2PTunnel (meshDevices[myNodeID][i], packet, protocol, src, dst);
                }
            }
        }
    }
  return true;
}

// --- 3. Split-Horizon Ingress (From ns-3 Mesh) ---
bool Ingress_From_Mesh (uint32_t myNodeID, Ptr<NetDevice> rxDevice, Ptr<const Packet> packet, uint16_t protocol,
                        const Address &src, const Address &dst, NetDevice::PacketType packetType)
{
  Ptr<Packet> pktCopy = packet->Copy();
  EthernetHeader eth;
  pktCopy->RemoveHeader (eth);
  
  // SPLIT HORIZON RULE: Packets arriving from the mesh are NEVER forwarded back into the mesh.
  // They are only sent UP to the physical Raspberry Pi attached to this node. 
  // This completely eliminates Layer 2 broadcast loops without needing STP.
  g_fdDev[myNodeID]->Send(pktCopy, eth.GetDestination(), eth.GetLengthType());
  
  return true;
}

// --- 4. Scheduled Apply Function (Strict Directional Mapping) ---
struct LinkChange {
  uint32_t src;
  uint32_t dst;
  double bwMbps;
  double dropRate;
};

void ApplyLinkChanges (std::vector<LinkChange> changes)
{
  NS_LOG_UNCOND ("\n--- Sim Time: " << Simulator::Now ().GetSeconds () << "s | Applying YAML Topology Updates ---");
  for (const auto& change : changes)
    {
      if (change.src < numNodes && change.dst < numNodes && change.src != change.dst)
        {
          double bw = change.bwMbps <= 0.0 ? 0.000001 : change.bwMbps;
          double drop = change.bwMbps <= 0.0 ? 1.0 : change.dropRate;

          DataRate newRate(bw * 1000000);
          
          // Create error model for the Destination node receiving from Source
          Ptr<RateErrorModel> emDst = CreateObject<RateErrorModel> ();
          emDst->SetAttribute ("ErrorRate", DoubleValue (drop));
          emDst->SetUnit (RateErrorModel::ERROR_UNIT_PACKET);

          // Create error model for the Source node receiving from Destination
          Ptr<RateErrorModel> emSrc = CreateObject<RateErrorModel> ();
          emSrc->SetAttribute ("ErrorRate", DoubleValue (drop));
          emSrc->SetUnit (RateErrorModel::ERROR_UNIT_PACKET);

          // --- FIX: Apply updates symmetrically to both sides of the P2P pipe ---
          
          // 1. Throttle the Forward Path (Src -> Dst)
          meshDevices[change.src][change.dst]->SetAttribute ("DataRate", DataRateValue (newRate));
          meshDevices[change.dst][change.src]->SetAttribute ("ReceiveErrorModel", PointerValue (emDst));
          
          // 2. Throttle the Return Path (Dst -> Src)
          meshDevices[change.dst][change.src]->SetAttribute ("DataRate", DataRateValue (newRate));
          meshDevices[change.src][change.dst]->SetAttribute ("ReceiveErrorModel", PointerValue (emSrc));

          NS_LOG_UNCOND ("  [Link Updated Symmetrically] " << (change.src + 1) << " <-> " << (change.dst + 1) 
                         << " | BW: " << bw << " Mbps | Drop Rate: " << drop);
        }
    }
}

// --- 5. Main Network Setup ---
int main (int argc, char *argv[])
{
  CommandLine cmd;
  cmd.Parse (argc, argv);

  GlobalValue::Bind ("SimulatorImplementationType", StringValue ("ns3::RealtimeSimulatorImpl"));

  NodeContainer meshNodes;
  meshNodes.Create (numNodes);

  // Initialize tracking matrix
  for(int i=0; i<numNodes; i++)
    for(int j=0; j<numNodes; j++)
      meshDevices[i][j] = nullptr;

  PointToPointHelper p2p;
  p2p.SetDeviceAttribute ("DataRate", StringValue ("100Mbps")); 
  p2p.SetChannelAttribute ("Delay", StringValue ("0ms"));
  p2p.SetQueue ("ns3::DropTailQueue<Packet>", "MaxSize", QueueSizeValue (QueueSize ("5000p")));

  // Build the Full Mesh (Creates N*(N-1)/2 Links)
  for (uint32_t i = 0; i < numNodes; ++i)
    {
      for (uint32_t j = i + 1; j < numNodes; ++j)
        {
          NodeContainer linkNodes (meshNodes.Get (i), meshNodes.Get (j));
          NetDeviceContainer devs = p2p.Install (linkNodes);

          Ptr<PointToPointNetDevice> devI = DynamicCast<PointToPointNetDevice> (devs.Get (0));
          Ptr<PointToPointNetDevice> devJ = DynamicCast<PointToPointNetDevice> (devs.Get (1));

          meshDevices[i][j] = devI; // Device on 'i' transmitting to 'j'
          meshDevices[j][i] = devJ; // Device on 'j' transmitting to 'i'

          // Bind the receiving callbacks (passing the local Node ID)
          devI->SetPromiscReceiveCallback (MakeBoundCallback (&Ingress_From_Mesh, i));
          devJ->SetPromiscReceiveCallback (MakeBoundCallback (&Ingress_From_Mesh, j));
        }
    }

  EmuFdNetDeviceHelper emuHelper;
  emuHelper.SetAttribute ("EncapsulationMode", StringValue ("Dix"));
  std::vector<std::string> vlanMapping = {"vlan101", "vlan102", "vlan103", "vlan104"};

  for (int i = 0; i < numNodes; ++i)
    {
      emuHelper.SetDeviceName (vlanMapping[i]);
      
      NetDeviceContainer devSide = emuHelper.Install (meshNodes.Get(i));
      g_fdDev[i] = devSide.Get(0);
      g_fdDev[i]->SetAttribute("Address", Mac48AddressValue(Mac48Address::Allocate()));

      // Bind the Pi ingest callback
      g_fdDev[i]->SetPromiscReceiveCallback (MakeBoundCallback (&Ingress_From_Vlan, i));
    }

  // --- Parse YAML File ---
  std::ifstream traceFile ("/home/ijoldenb/ns-3.48/scratch/topology_trace.yaml");
  if (!traceFile.is_open ())
    {
      NS_FATAL_ERROR ("Could not open topology_trace.yaml!");
    }
  
  std::string line;
  double scheduleTime = 0.0;
  bool inLink = false;
  LinkChange tempChange;
  std::map<double, std::vector<LinkChange>> timelineMap;

  while (std::getline (traceFile, line))
    {
      if (line.empty() || line.find (":") == std::string::npos) continue;

      try 
        {
          std::string valStr = line.substr (line.find (":") + 1);

          if (line.find ("- time:") != std::string::npos)
            {
              scheduleTime = std::stod (valStr);
            }
          else if (line.find ("- src:") != std::string::npos)
            {
              // Subtract 1 to map YAML Node 1-4 to internal Node 0-3
              tempChange.src = std::stoi (valStr) - 1; 
              inLink = true;
            }
          else if (inLink && line.find ("dst:") != std::string::npos)
            {
              // Subtract 1 to map YAML Node 1-4 to internal Node 0-3
              tempChange.dst = std::stoi (valStr) - 1; 
            }
          else if (inLink && line.find ("bw:") != std::string::npos)
            {
              tempChange.bwMbps = std::stod (valStr);
            }
          else if (inLink && line.find ("drop:") != std::string::npos)
            {
              tempChange.dropRate = std::stod (valStr);
              
              timelineMap[scheduleTime].push_back (tempChange);
              inLink = false;
              tempChange = LinkChange(); // Reset to prevent bleeding
            }
        } 
      catch (const std::exception& e) 
        {
          NS_FATAL_ERROR ("\n[YAML PARSE ERROR] Crash on line: '" << line << "'\nException: " << e.what ());
        }
    }

  for (auto const& entry : timelineMap)
    {
      Simulator::Schedule (Seconds (entry.first), &ApplyLinkChanges, entry.second);
    }

  Time stopTime = Seconds (3600.0); 
  Simulator::Stop (stopTime);
  Simulator::Schedule (stopTime - Seconds(1.0), &KeepAliveDummyEvent);

  NS_LOG_UNCOND ("================================================================");
  NS_LOG_UNCOND ("ns-3 Dynamic PointToPoint FULL MESH Active (L2 Split-Horizon)");
  NS_LOG_UNCOND ("================================================================");

  Simulator::Run ();
  Simulator::Destroy ();
  
  return 0;
}

File: scratch/4sat-link-emu-multi-dynamic-p2p-graph.py
import sys
from ns import ns
import cppyy

# ==============================================================================
# --- 0. CRITICAL C++ TO PYTHON CALLBACK BRIDGE & CASTING ENGINE ---
# ==============================================================================
cppyy.cppdef("""
#include "ns3/net-device.h"
#include "ns3/packet.h"
#include "ns3/callback.h"
#include "ns3/simulator.h"
#include "ns3/event-impl.h"
#include "ns3/mac48-address.h"
#include "ns3/address.h"
#include <functional>
#include <vector>
#include <memory>

namespace ns3 {

class PromiscCallbackBridge {
public:
    std::function<bool(Ptr<NetDevice>, Ptr<const Packet>, uint16_t, const Address&, const Address&, NetDevice::PacketType)> m_pyFunc;

    PromiscCallbackBridge(std::function<bool(Ptr<NetDevice>, Ptr<const Packet>, uint16_t, const Address&, const Address&, NetDevice::PacketType)> pyFunc)
        : m_pyFunc(pyFunc) {}

    bool Invoke(Ptr<NetDevice> dev, Ptr<const Packet> pkt, uint16_t proto, const Address& src, const Address& dst, NetDevice::PacketType type) {
        return m_pyFunc(dev, pkt, proto, src, dst, type);
    }
};

static std::vector<std::shared_ptr<PromiscCallbackBridge>> g_callbackBridges;

Callback<bool, Ptr<NetDevice>, Ptr<const Packet>, uint16_t, const Address&, const Address&, NetDevice::PacketType>
CreatePromiscCallback(std::function<bool(Ptr<NetDevice>, Ptr<const Packet>, uint16_t, const Address&, const Address&, NetDevice::PacketType)> pyFunc) {
    auto bridge = std::make_shared<PromiscCallbackBridge>(pyFunc);
    g_callbackBridges.push_back(bridge);
    return MakeCallback(&PromiscCallbackBridge::Invoke, bridge.get());
}

class PythonEventImpl : public EventImpl {
public:
    std::function<void()> m_func;
    PythonEventImpl(std::function<void()> func) : m_func(func) {}
    virtual void Notify() override {
        m_func();
    }
};

void SchedulePythonEvent(Time delay, std::function<void()> func) {
    Simulator::Schedule(delay, Create<PythonEventImpl>(func));
}

// Native C++ helper to securely map Mac48Address to a base Address object
Address ToAddress(const Mac48Address& mac) {
    return Address(mac);
}

}

using ns3::CreatePromiscCallback;
using ns3::SchedulePythonEvent;
using ns3::ToAddress;
""")

# --- Global Network Infrastructure ---
numNodes = 4

# Mesh Tracking Matrix: meshDevices[SrcNode][DstNode]
meshDevices = [[None for _ in range(numNodes)] for _ in range(numNodes)]
g_fdDev = [None for _ in range(numNodes)]

# Global MAC Address Translation table for distributed L2 switching
macToNodeMap = {}

# Reference Vault to prevent Python GC cleanup
g_keepAlive = []


# --- LIVE TOPOLOGY REPORTING ENGINE ---
def PrintCurrentTopology():
    print("\n====================================================")
    print(f" CORE TOPOLOGY REPORT | Sim Time: {ns.Simulator.Now().GetSeconds()}s")
    print("====================================================")
    
    for i in range(numNodes):
        for j in range(i + 1, numNodes):
            if meshDevices[i][j] is None:
                continue
            
            # 1. Fetch data rate attribute directly from the active PointToPoint device
            bpsVal = ns.DataRateValue()
            meshDevices[i][j].GetAttribute("DataRate", bpsVal)
            bwMbps = bpsVal.Get().GetBitRate() / 1000000.0
            
            # 2. Extract the Receive Error Model via the dynamic pointer attribute wrapper
            ptrVal = ns.PointerValue()
            meshDevices[j][i].GetAttribute("ReceiveErrorModel", ptrVal)
            em = ptrVal.GetObject()  
            
            dropRate = 0.0
            if em:
                dropVal = ns.DoubleValue()
                em.GetAttribute("ErrorRate", dropVal)
                dropRate = dropVal.Get()
                
            print(f"  Node {i + 1} <-> Node {j + 1} | Bandwidth: {bwMbps:.2f} Mbps | Packet Loss: {dropRate * 100.0:.1f}%")
            
    print("====================================================\n")


# --- 1. Custom Encapsulation Tunnel ---
def SendOverP2PTunnel(dev, packet, protocol, src_addr, dst_addr):
    pktCopy = packet.Copy()
    eth = ns.EthernetHeader()
    eth.SetSource(ns.Mac48Address.ConvertFrom(src_addr))
    eth.SetDestination(ns.Mac48Address.ConvertFrom(dst_addr))
    eth.SetLengthType(protocol)
    
    pktCopy.AddHeader(eth)
    dev.Send(pktCopy, dev.GetBroadcast(), 0x0800)


# --- 2. Distributed Switch Ingress (From Physical Raspberry Pi) ---
def make_vlan_ingress_callback(myNodeID):
    def Ingress_From_Vlan(rxDevice, packet, protocol, src, dst, packetType):
        srcMac = ns.Mac48Address.ConvertFrom(src)
        dstMac = ns.Mac48Address.ConvertFrom(dst)

        if not srcMac.IsBroadcast():
            macToNodeMap[str(srcMac)] = myNodeID

        if dstMac.IsBroadcast() or dstMac.IsGroup():
            for i in range(numNodes):
                if i != myNodeID and meshDevices[myNodeID][i] is not None:
                    SendOverP2PTunnel(meshDevices[myNodeID][i], packet, protocol, src, dst)
        else: 
            dst_str = str(dstMac)
            if dst_str in macToNodeMap:
                targetNode = macToNodeMap[dst_str]
                if targetNode != myNodeID and meshDevices[myNodeID][targetNode] is not None:
                    SendOverP2PTunnel(meshDevices[myNodeID][targetNode], packet, protocol, src, dst)
            else:
                for i in range(numNodes):
                    if i != myNodeID and meshDevices[myNodeID][i] is not None:
                        SendOverP2PTunnel(meshDevices[myNodeID][i], packet, protocol, src, dst)
        return True
    return Ingress_From_Vlan


# --- 3. Split-Horizon Ingress (From Inter-Satellite Mesh Links) ---
def make_mesh_ingress_callback(myNodeID):
    def Ingress_From_Mesh(rxDevice, packet, protocol, src, dst, packetType):
        pktCopy = packet.Copy()
        eth = ns.EthernetHeader()
        pktCopy.RemoveHeader(eth)
        
        # Invokes the custom C++ bridge casting function to resolve the signature mapping crash
        g_fdDev[myNodeID].Send(pktCopy, cppyy.gbl.ToAddress(eth.GetDestination()), eth.GetLengthType())
        
        return True
    return Ingress_From_Mesh


# --- 4. Symmetric Dynamic Topology Engine ---
def ApplyLinkChanges(changes):
    for change in changes:
        src = change['src']
        dst = change['dst']
        
        if src < numNodes and dst < numNodes and src != dst:
            bw = 0.000001 if change['bwMbps'] <= 0.0 else change['bwMbps']
            drop = 1.0 if change['bwMbps'] <= 0.0 else change['dropRate']

            newRate = ns.DataRate(f"{bw}Mbps")
            
            emDst = ns.CreateObject[ns.RateErrorModel]()
            emDst.SetAttribute("ErrorRate", ns.DoubleValue(drop))
            emDst.SetUnit(ns.RateErrorModel.ERROR_UNIT_PACKET)

            emSrc = ns.CreateObject[ns.RateErrorModel]()
            emSrc.SetAttribute("ErrorRate", ns.DoubleValue(drop))
            emSrc.SetUnit(ns.RateErrorModel.ERROR_UNIT_PACKET)

            meshDevices[src][dst].SetAttribute("DataRate", ns.DataRateValue(newRate))
            meshDevices[dst][src].SetAttribute("ReceiveErrorModel", ns.PointerValue(emDst))
            
            meshDevices[dst][src].SetAttribute("DataRate", ns.DataRateValue(newRate))
            meshDevices[src][dst].SetAttribute("ReceiveErrorModel", ns.PointerValue(emSrc))

    PrintCurrentTopology()


def KeepAliveDummyEvent():
    pass


# --- 5. Dedicated YAML Parsing Engine ---
def parse_topology_trace(file_path):
    """
    Reads the YAML topology trace and structures it into a dictionary mapped by simulation time.
    """
    timelineMap = {}
    scheduleTime = 0.0
    inLink = False
    tempChange = {}

    try:
        with open(file_path, "r") as traceFile:
            for line in traceFile:
                line = line.strip()
                if not line or ":" not in line:
                    continue
                
                key, valStr = [x.strip() for x in line.split(":", 1)]

                if "- time" in key:
                    scheduleTime = float(valStr)
                elif "- src" in key:
                    # YAML is 1-indexed, Python arrays are 0-indexed
                    tempChange['src'] = int(valStr) - 1
                    inLink = True
                elif inLink and "dst" in key:
                    tempChange['dst'] = int(valStr) - 1
                elif inLink and "bw" in key:
                    tempChange['bwMbps'] = float(valStr)
                elif inLink and "drop" in key:
                    tempChange['dropRate'] = float(valStr)
                elif inLink and "latency" in key:
                    tempChange['latency'] = float(valStr)
                    
                    # Latency is the last element in the YAML block, trigger the append here
                    if scheduleTime not in timelineMap:
                        timelineMap[scheduleTime] = []
                    timelineMap[scheduleTime].append(tempChange.copy())
                    
                    inLink = False
                    tempChange = {}
    except FileNotFoundError:
        print(f"FATAL ERROR: Missing trajectory map at {file_path}!")
        sys.exit(1)
        
    return timelineMap


# --- 6. Real-Time Network Execution Environment ---
def main():
    ns.CommandLine().Parse(sys.argv)
    ns.GlobalValue.Bind("SimulatorImplementationType", ns.StringValue("ns3::RealtimeSimulatorImpl"))

    meshNodes = ns.NodeContainer()
    meshNodes.Create(numNodes)

    p2p = ns.PointToPointHelper()
    p2p.SetDeviceAttribute("DataRate", ns.StringValue("100Mbps"))
    p2p.SetChannelAttribute("Delay", ns.StringValue("0ms"))
    p2p.SetDeviceAttribute("Mtu", ns.UintegerValue(1550))
    p2p.SetQueue("ns3::DropTailQueue<Packet>", "MaxSize", ns.QueueSizeValue(ns.QueueSize("5000p")))

    for i in range(numNodes):
        for j in range(i + 1, numNodes):
            linkNodes = ns.NodeContainer()
            linkNodes.Add(meshNodes.Get(i))
            linkNodes.Add(meshNodes.Get(j))
            
            devs = p2p.Install(linkNodes)
            devI = devs.Get(0)
            devJ = devs.Get(1)

            meshDevices[i][j] = devI
            meshDevices[j][i] = devJ

            cbI = make_mesh_ingress_callback(i)
            cbJ = make_mesh_ingress_callback(j)
            g_keepAlive.append(cbI)
            g_keepAlive.append(cbJ)

            devI.SetPromiscReceiveCallback(cppyy.gbl.CreatePromiscCallback(cbI))
            devJ.SetPromiscReceiveCallback(cppyy.gbl.CreatePromiscCallback(cbJ))

    emuHelper = ns.EmuFdNetDeviceHelper()
    emuHelper.SetAttribute("EncapsulationMode", ns.StringValue("Dix"))
    vlanMapping = ["vlan101", "vlan102", "vlan103", "vlan104"]

    for i in range(numNodes):
        emuHelper.SetDeviceName(vlanMapping[i])
        devSide = emuHelper.Install(meshNodes.Get(i))
        g_fdDev[i] = devSide.Get(0)
        
        mac_addr = ns.Mac48Address.Allocate()
        g_fdDev[i].SetAttribute("Address", ns.Mac48AddressValue(mac_addr))

        cbVlan = make_vlan_ingress_callback(i)
        g_keepAlive.append(cbVlan)
        g_fdDev[i].SetPromiscReceiveCallback(cppyy.gbl.CreatePromiscCallback(cbVlan))

    # --- Call the standalone parser ---
    trace_file_path = "/home/ijoldenb/ns-3.48/scratch/topology_trace.yaml"
    timelineMap = parse_topology_trace(trace_file_path)

    # Schedule the link changes
    for t_time, changes in timelineMap.items():
        evt = lambda c=changes: ApplyLinkChanges(c)
        g_keepAlive.append(evt)
        ns.SchedulePythonEvent(ns.Seconds(t_time), evt)

    stopTime = ns.Seconds(3600.0)
    ns.Simulator.Stop(stopTime)
    ns.SchedulePythonEvent(stopTime - ns.Seconds(1.0), KeepAliveDummyEvent)

    print("================================================================")
    print("ns-3 Dynamic PointToPoint FULL MESH Active (Python L2 Split-Horizon)")
    print("================================================================")

    ns.Simulator.Run()
    ns.Simulator.Destroy()


if __name__ == '__main__':
    main()